# Importing libraries 

In [9]:
#Importing Libraries 
import pandas as pd
import numpy as np
import os 

path = r'C:\Users\ANITA BOADU\Achievement 4'

# Import the dataframe ords_prods_merge as 
ords_prods_merge = pd.read_pickle(os.path.join(path,'Data','Prepared Data','ords_prods_merge.pkl'))

# Create the first 1,000,000-row subset
df = ords_prods_merge[:1000000]

# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'ords_prods_merge_new_column.pkl'))




In [11]:
df.shape

(1000000, 16)

In [12]:
#define a function 

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High-range product'
  else: return 'Not enough data'

#######
Low-range product → when the price is less than or equal to 5
Mid-range product → when the price is greater than 5 and less than or equal to 15
High-range product → when the price is greater than 15
Not enough data → only used if none of the above conditions apply (e.g., missing or unexpected values)

In [13]:
print(df.columns.tolist())


['Unnamed: 0', 'order_id', 'user_id', 'eval_set', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order', 'product_id', 'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id', 'department_id', 'prices']


In [14]:
#Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\ANITA BOADU\AppData\Local\Temp\ipykernel_18032\3482686425.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


########
df['price_range'] = df.apply(price_label, axis=1) means that Python uses the price_label function on every row of the dataframe to decide whether the product is low-, mid-, or high-range, and then saves that label in a new column called price_range.

In [15]:
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product     672143
Low-range product     314646
High-range product     12292
Not enough data          919
Name: count, dtype: int64

#######df['price_range'].value_counts(dropna=False) counts how many times each price label appears in the price_range column and also includes any missing values in the count.his result shows how many products fall into each price category:

672,143 are mid-range products
314,646 are low-range products
12,292 are high-range products
919 rows didn’t have enough data to assign a category.

In [7]:
df['prices'].max()

99999.0

In [10]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [9]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product     672143
Low-range product     314646
High-range product     12292
Not enough data          919
Name: count, dtype: int64

In [21]:
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif row['prices'] <= 15:
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range product'
    else:
        return 'Not enough data'

# Apply to the full dataframe
ords_prods_merge['price_range'] = ords_prods_merge.apply(price_label, axis=1)


In [23]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [24]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [25]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [26]:
ords_prods_merge['price_range'].value_counts(dropna=False)

price_range
Mid-range product     20462144
Low-range product      9476774
High-range product      389845
Not enough data          28194
Name: count, dtype: int64

In [30]:
print(ords_prods_merge.columns)

Index(['Unnamed: 0', 'order_id', 'user_id', 'eval_set', 'order_number',
       'order_dow', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'price_range'],
      dtype='object')


In [32]:
ords_prods_merge['order_dow'].value_counts(dropna=False)

order_dow
0    5784308
1    5309089
6    4194576
5    3955888
2    3951349
3    3603962
4    3557785
Name: count, dtype: int64

In [34]:
result = []

for value in ords_prods_merge["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [35]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Reg

In [36]:
ords_prods_merge['busiest_day']=result

In [37]:
ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    21014864
Busiest day        5784308
Least busy         3557785
Name: count, dtype: int64

In [38]:
# Two busiest days (highest counts)
busiest_days = day_counts.nlargest(2).index.tolist()

# Two slowest days (lowest counts)
slowest_days = day_counts.nsmallest(2).index.tolist()

print("Busiest days:", busiest_days)
print("Slowest days:", slowest_days)


Busiest days: [0, 1]
Slowest days: [4, 3]


In [39]:
# Prepare a list to store the labels
day_labels = []

# Loop through each value in 'order_dow'
for day in ords_prods_merge['order_dow']:
    if day in busiest_days:
        label = "Busiest days"
    elif day in slowest_days:
        label = "Slowest days"
    else:
        label = "Moderate day"
    day_labels.append(label)

# Add this list as a new column in the DataFrame
ords_prods_merge['day_status_updated'] = day_labels


In [40]:
ords_prods_merge['day_status_updated'].value_counts()

day_status_updated
Moderate day    12101813
Busiest days    11093397
Slowest days     7161747
Name: count, dtype: int64

########## Observations on `day_status_updated` column

- The column correctly labels the **two busiest days** as `"Busiest days"`.
- The column correctly labels the **two slowest days** as `"Slowest days"`.
- All other days are labeled `"Moderate day"`.
- The distribution of orders matches the original `order_dow` frequencies, confirming that the labels accurately reflect busyness.
- This column provides a clear summary of daily activity, making it easier to identify peak and slow periods for stakeholders.


In [41]:
# Count the number of orders per hour
hour_counts = ords_prods_merge['order_hour_of_day'].value_counts()
print(hour_counts)


order_hour_of_day
10    2596274
11    2566940
14    2519533
15    2489838
13    2489731
12    2448050
16    2367121
9     2313739
17    1945590
8     1624168
18    1522301
19    1170216
20     910845
7      845501
21     746953
22     592918
23     376174
6      275087
0      203627
1      108193
5       82805
2       64017
4       49442
3       47894
Name: count, dtype: int64


In [43]:
hour_counts = ords_prods_merge['order_hour_of_day'].value_counts()


In [44]:
# Top 25% of hours → Most orders
most_orders_threshold = hour_counts.quantile(0.75)

# Bottom 25% of hours → Fewest orders
fewest_orders_threshold = hour_counts.quantile(0.25)


In [45]:
period_labels = []

for hour in ords_prods_merge['order_hour_of_day']:
    count = hour_counts[hour]  # Number of orders for this hour
    
    if count >= most_orders_threshold:
        label = "Most orders"
    elif count <= fewest_orders_threshold:
        label = "Fewest orders"
    else:
        label = "Average orders"
    
    period_labels.append(label)

# Add the labels as a new column
ords_prods_merge['busiest_period_of_day'] = period_labels


######### Creating the `busiest_period_of_day` Column

The goal of this code is to categorize each order in the `ords_prods_merge` DataFrame according to the number of orders in its corresponding hour of the day. We create a new column, `busiest_period_of_day`, with three possible labels: `"Most orders"`, `"Average orders"`, and `"Fewest orders"`.

######## Steps Explained:

1. **Create an empty list to store labels**
```python
period_labels = []


In [46]:
ords_prods_merge['busiest_period_of_day'].value_counts()


busiest_period_of_day
Most orders       15110366
Average orders    14690613
Fewest orders       555978
Name: count, dtype: int64

In [47]:
# Print the frequency of each label in the new column
print(ords_prods_merge['busiest_period_of_day'].value_counts())


busiest_period_of_day
Most orders       15110366
Average orders    14690613
Fewest orders       555978
Name: count, dtype: int64


########## Observations on `busiest_period_of_day` Column

- The new column categorizes each order hour into three labels:  
  1. **Most orders** – hours with the highest number of orders  
  2. **Average orders** – hours with moderate order volume  
  3. **Fewest orders** – hours with the lowest number of orders
  
  ####### Observations

- The majority of orders occur during **Most orders** and **Average orders** hours, indicating that a large portion of activity happens during peak and moderate periods.  
- **Fewest orders** hours are relatively rare, showing that certain hours of the day have very low app usage.  